src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -872,39 +900,67 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message,
 
                 /* verify that the requested address is from the pool, and either
                    owned by the current client or free */
-                if (pool_offset >= 0 &&
-                    server->bound_leases[pool_offset] == existing_lease) {
-                        DHCPLease *lease;
-                        usec_t time_now = 0;
-
-                        if (!existing_lease) {
-                                lease = new0(DHCPLease, 1);
-                                if (!lease)
-                                        return -ENOMEM;
-                                lease->address = address;
-                                lease->client_id.data = memdup(req->client_id.data,
-                                                               req->client_id.length);
-                                if (!lease->client_id.data) {
-                                        free(lease);
-                                        return -ENOMEM;
-                                }
-                                lease->client_id.length = req->client_id.length;
-                                memcpy(&lease->chaddr, &req->message->chaddr,
-                                       ETH_ALEN);
-                                lease->gateway = req->message->giaddr;
-                        } else
-                                lease = existing_lease;
+                if (static_lease) {
+                        _cleanup_free_ DHCPLease *lease;
+                        usec_t time_now = 0, expiration = 0;
 
                         r = sd_event_now(server->event,
                                          clock_boottime_or_monotonic(),
                                          &time_now);
+                        if (r < 0)
+                                return r;
+
+                        expiration = req->lifetime * USEC_PER_SEC + time_now;
+
+                        r = prepare_new_lease(&lease, static_lease->address, &req->client_id,
+                                              req->message->chaddr, req->message->giaddr, expiration);
+                        if (r < 0)
+                                return r;
+
+                        r = server_send_ack(server, req, address);
                         if (r < 0) {
-                                if (!existing_lease)
-                                        dhcp_lease_free(lease);
+                                /* this only fails on critical errors */
+                                log_dhcp_server_errno(server, r, "Could not send ack: %m");
+
                                 return r;
```

Comment:
log_dhcp_server_errno() returns the error you pass in, hence you can combine these two lines into one:

```c
return log_dhcp_server_errno(server, r, …);
```